File: app/src/main/java/ru/ulstu_team/ulstuschedule/ui/cathedries/CathedriesPresenter.java
package ru.ulstu_team.ulstuschedule.ui.cathedries;

import java.util.List;

import javax.inject.Inject;

import io.realm.RealmQuery;
import ru.ulstu_team.ulstuschedule.data.DataManager;
import ru.ulstu_team.ulstuschedule.data.model.Cathedra;
import ru.ulstu_team.ulstuschedule.data.remote.Schedule;
import ru.ulstu_team.ulstuschedule.data.remote.ScheduleRequest;
import ru.ulstu_team.ulstuschedule.ui.base.BasePresenter;

public class CathedriesPresenter extends BasePresenter<CathedriesMvpView> {

    private DataManager mDataManager;

    @Inject
    public CathedriesPresenter(DataManager dataManager) {
        mDataManager = dataManager;
    }

    public void loadCathedries() {
        checkViewAttached();

        List<Cathedra> cathedries = getRealmQuery().findAll();
        if (!cathedries.isEmpty()) {
            getMvpView().showCathedries(cathedries);
        } else {
            mDataManager.executeRequest(getRequest());
        }
    }

    private RealmQuery<Cathedra> getRealmQuery() {
        return mRealm.where(Cathedra.class);
    }

    private ScheduleRequest getRequest() {
        return new ScheduleRequest(Schedule.CATHEDRIES, Cathedra.class, getRealmQuery(),
                new ScheduleRequest.Callbacks() {
                    @Override
                    public void onSuccess() {
                        List<Cathedra> cathedries = getRealmQuery().findAll();
                        getMvpView().showCathedries(cathedries);
                    }

                    @Override
                    public void onError(Exception e) {
                        getMvpView().showError();
                    }
                });
    }
}

File: app/src/main/java/ru/ulstu_team/ulstuschedule/ui/cathedries/CathedriesPresenter.kt
package ru.ulstu_team.ulstuschedule.ui.cathedries

import ru.ulstu_team.ulstuschedule.data.DataManager
import ru.ulstu_team.ulstuschedule.data.remote.RequestCallbacks
import ru.ulstu_team.ulstuschedule.ui.base.BasePresenter
import javax.inject.Inject

class CathedriesPresenter
@Inject
constructor(private val mDataManager: DataManager) : BasePresenter<CathedriesMvpView>() {

    fun loadCathedries() {
        checkViewAttached()

        val cathedries = mDataManager.getCathedries()
        if (cathedries.isNotEmpty())
            mvpView!!.showCathedries(cathedries)
        else
            mDataManager.loadCathedries(getCathedriesRequestCallbacks())
    }

    fun loadCathedries(facultyId: Int) {
        checkViewAttached()

        val cathedries = mDataManager.getFacultyCathedries(facultyId)
        if (!cathedries.isEmpty())
            mvpView!!.showCathedries(cathedries)
        else
            mDataManager.loadFacultyCathedries(facultyId,
                    getFacultyCathedriesRequestCallbacks(facultyId))
    }

    private fun getCathedriesRequestCallbacks(): RequestCallbacks =
            object: RequestCallbacks {
                override fun onSuccess() {
                    val cathedries = mDataManager.getCathedries()
                    if (cathedries.isNotEmpty())
                        mvpView!!.showCathedries(cathedries)
                    else
                        mvpView!!.showEmptyList()
                }

                override fun onError(e: Exception) {
                    mvpView!!.showError()
                }
            }

    private fun getFacultyCathedriesRequestCallbacks(facultyId: Int): RequestCallbacks =
            object: RequestCallbacks {
                override fun onSuccess() {
                    val cathedries = mDataManager.getFacultyCathedries(facultyId)
                    if (cathedries.isNotEmpty())
                        mvpView!!.showCathedries(cathedries)
                    else
                        mvpView!!.showEmptyList()
                }

                override fun onError(e: Exception) {
                    mvpView!!.showError()
                }
            }
}
